### PR TITLE
Fix GPU rendering on Linux

### LINKED ISSUE #6480: Add support for Linux-based systems to GPU plugin
Currently GPU plugin do not works on Linux systems. The problem is that even after adding proper JOGL natives for Linux, the input on Linux is broken (confirmed by 2 separate people trying to run GPU plugin on Linux). Possible solution is switch from JOGL to LWJGL2 what works just fine or to LWJGL3 (untested, we would need to use their experimental JAWT api what they do not have documented very well).

EDIT: Problem with LWJGL2 is that it do not propagates any input to canvas, so it needs to be proxies. JOGL simply grabs all input and does nothing with it on Linux.

### PR DESCRIPTION
**Pull request by whitehooder**
Fixes #6480.

### CODE REVIEW

**Comment by doorknob60** [comment]
The GPU plugin is working flawlessly for me on GTX 1070 in Arch with this change. For what it's worth I do have a manually compiled copy of jogl from git in my natives folder, though. Am not sure what building it from scratch would put in there.

**Comment by gatlee** [comment]
`WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.jogamp.common.os.NativeLibrary$3 (file:/home/gatlee/.m2/repository/org/jogamp/gluegen/gluegen-rt/2.3.2/gluegen-rt-2.3.2.jar) to method java.lang.ClassLoader.findLibrary(java.lang.String)
WARNING: Please consider reporting this to the maintainers of com.jogamp.common.os.NativeLibrary$3
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
2018-11-17 19:59:30 [Thread-2] ERROR n.r.client.plugins.gpu.GpuPlugin - Error starting GPU plugin
com.jogamp.opengl.GLException: Profile GL4bc is not available on X11GraphicsDevice[type .x11, connection :0, unitID 0, handle 0x0, owner false, ResourceToolkitLock[obj 0xb902752, isOwner false, <382b2181, 680e8151>[count 0, qsz 0, owner <NULL>]]], but: [GLProfile[GLES1/GLES1.hw], GLProfile[GLES2/GLES3.hw], GLProfile[GL2ES1/GL2.hw], GLProfile[GL4ES3/GL4.hw], GLProfile[GL2ES2/GL4.hw], GLProfile[GL2/GL2.hw], GLProfile[GL2/GL2.hw], GLProfile[GLES3/GLES3.hw], GLProfile[GL4/GL4.hw], GLProfile[GL3/GL4.hw], GLProfile[GL2GL3/GL4.hw]]
	at com.jogamp.opengl.GLProfile.get(GLProfile.java:991)
	at com.jogamp.opengl.GLProfile.get(GLProfile.java:1004)
	at net.runelite.client.plugins.gpu.GpuPlugin.lambda$startUp$0(GpuPlugin.java:230)
	at net.runelite.client.callback.ClientThread.lambda$invoke$0(ClientThread.java:48)
	at net.runelite.client.callback.ClientThread.invoke(ClientThread.java:99)
	at net.runelite.client.callback.Hooks.clientMainLoop(Hooks.java:152)
	at client.methodDraw(client.java:1246)
	at client.ax(client.java)
	at ah.am(ah.java:364)
	at ah.run(ah.java:318)
	at java.base/java.lang.Thread.run(Thread.java:844)
2018-11-17 19:59:30 [Thread-2] WARN  n.r.client.callback.ClientThread - Exception in invoke
java.lang.NullPointerException: null
	at net.runelite.client.plugins.gpu.GLUtil.glDeleteTexture(GLUtil.java:106)
	at net.runelite.client.plugins.gpu.TextureManager.freeTextureArray(TextureManager.java:79)
	at net.runelite.client.plugins.gpu.GpuPlugin.lambda$shutDown$1(GpuPlugin.java:313)
	at net.runelite.client.callback.ClientThread.lambda$invoke$0(ClientThread.java:48)
	at net.runelite.client.callback.ClientThread.invoke(ClientThread.java:61)
	at net.runelite.client.callback.ClientThread.invoke(ClientThread.java:46)
	at net.runelite.client.plugins.gpu.GpuPlugin.shutDown(GpuPlugin.java:309)
	at net.runelite.client.plugins.gpu.GpuPlugin.lambda$startUp$0(GpuPlugin.java:300)
	at net.runelite.client.callback.ClientThread.lambda$invoke$0(ClientThread.java:48)
	at net.runelite.client.callback.ClientThread.invoke(ClientThread.java:99)
	at net.runelite.client.callback.Hooks.clientMainLoop(Hooks.java:152)
	at client.methodDraw(client.java:1246)
	at client.ax(client.java)
	at ah.am(ah.java:364)
	at ah.run(ah.java:318)
	at java.base/java.lang.Thread.run(Thread.java:844)
2018-11-17 19:59:30 [Thread-2] WARN  n.r.client.callback.ClientThread - Exception in invoke
java.lang.NullPointerException: null
	at net.runelite.client.plugins.gpu.GLUtil.glDeleteTexture(GLUtil.java:106)
	at net.runelite.client.plugins.gpu.TextureManager.freeTextureArray(TextureManager.java:79)
	at net.runelite.client.plugins.gpu.GpuPlugin.lambda$shutDown$1(GpuPlugin.java:313)
	at net.runelite.client.callback.ClientThread.lambda$invoke$0(ClientThread.java:48)
	at net.runelite.client.callback.ClientThread.invoke(ClientThread.java:99)
	at net.runelite.client.callback.Hooks.clientMainLoop(Hooks.java:152)
	at client.methodDraw(client.java:1246)
	at client.ax(client.java)
	at ah.am(ah.java:364)
	at ah.run(ah.java:318)
	at java.base/java.lang.Thread.run(Thread.java:844)`

This error is given on my end when attempting to turn on the GPU rendering. Not sure if the issue is on my end.

**Comment by holmsten** [comment]
This fix works well for me building from scratch. I'm using a GTX 770 (driver 384.13) on Linux Mint 18.3.

**Comment by faultyme** [comment]
Works on Gentoo Linux with Vega 64(Mesa 18.3.0-rc3).
Thank you :smile:

**Comment by MP2E** [comment]
Works perfectly on latest NixOS unstable x86_64 with nvidia binary drivers, thank you!

**Comment by Desetude** [comment]
Works great on Arch Linux with GTX 970 & nvidia binary driver 410.66.

**Comment by whitehooder** [comment]
@gatlee Illegal reflective access comes from running Java version > 8, but isn't what's causing the GPU plugin to fail. In your case I don't think there's a way with Intel Graphics. The problem is that the drivers don't support OpenGL 4.3 and maybe the hardware doesn't either.

A good addition though would be an error message explaining why it fails to enable and potential things to try such as a proprietary driver.

**Comment by whitehooder** [comment]
I'm not completely sure. Didn't try removing it after adding the glContext.destroy(). The thing it's supposed to fix is a warning about the jawtWindow surface not being locked when reenabling the plugin.

**Comment by pacf531** [comment]
@whitehooder I don't think that is the reason why the GPU plugin is failing to load on Intel integrated graphics, as I got the same error and stack trace but I have a Skylake processor with a Intel HD Graphics 510 which should support OpenGL 4 so this shouldn't be happening naturally. As it turns out, what I think is happening is that GL4bc, as hinted in its name, most likely needs a compatibility context provided by the OpenGL implementation as opposed to the core context which is usually what is used.

If I do a glxinfo command to query OpenGL information on my Ubuntu system, the following input snippet is shown:

```
...
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Intel Open Source Technology Center (0x8086)
    Device: Mesa DRI Intel(R) HD Graphics 510 (Skylake GT1)  (0x1906)
    Version: 18.2.4
    Accelerated: yes
    Video memory: 3072MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.5
    Max compat profile version: 3.0
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
...
```

As you can see, the core profile reports correctly with a 4.5 but the compat profile only goes to 3.0. JOGL probably only looks at the 3.0 and concludes your compat profile version is not high enough as opposed to your core profile version and thus using GL4bc fails. I manually changed GPUPlugin.java to use GL4 instead for the profile and the GPU plugin can be toggled without errors and will stay on, but I just get a black screen so not sure if there is something else that is failing.

**Comment by deathbeam** [comment]
Yeah, solution to your problem is simply using core profile, what is easy enough, but probably should be part of separate PR.

**Comment by matt2134711** [comment]
Working flawlessly on my system (w/ nvidia-driver 390.67). Thank you so much!

**Comment by pacf531** [comment]
The override doesn't work, you'll get the same errors as before. I tried using "4.3COMPAT" as well as using "4.5" and "4.5COMPAT" with the same results.

**Comment by deathbeam** [comment]
This PR broke support on windows, so updated it to not do that, can someone retest this now after my changes?

**Comment by deathbeam** [comment]
Actually I tested it now on both Linux and Windows and its fine

**Comment by gatlee** [comment]
Can confirm that changing the environment variable doesn't fix the issue.

**Comment by gatlee** [comment]
> ...
Vendor: Intel Open Source Technology Center (0x8086)
    Device: Mesa DRI Intel(R) HD Graphics 5500 (Broadwell GT2)  (0x1616)
    Version: 18.2.4
    Accelerated: yes
    Video memory: 3072MB
    Unified memory: yes
    Preferred profile: core (0x1)
    Max core profile version: 4.3
    Max compat profile version: 4.3
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.1
...

Here is my glxinfo info too. My Max compat profile version is higher than @pacf531 's so it might not be due to that.

**Comment by deathbeam** [comment]
I created PR here https://github.com/runelite/runelite/pull/6534 that when built on top of this PR should solve your issue @gatlee

**Comment by whitehooder** [comment]
Thanks for testing & fixing it on Windows @deathbeam. I tried checking the lock, but didn't spend enough time to get around to that second check you added.

**Comment by Craio** [comment]
Just wanted to add this works on my machine too.

Linux-x86_64 4.19.2-arch1-1-ARCH
NVIDIA Quadro K2000M Driver Version: 410.73

Thanks for all the hard work.

**Comment by chiragbharadwaj** [comment]
I didn't see any comments about Ubuntu set-ups, so I can confirm that this works on my machine as well after a clone+rebuild of this live pull request.

OS: Ubuntu 18.04.1 LTS x86_64
Kernel: 4.15.0
CPU: AMD Ryzen 7 2700X (16) @ 3.7 GHz
GPU: NVIDIA GeForce GTX 1070
GPU Driver: 390.87

Thank you for all of the work you've put in. This is truly amazing! I have just one question -- if the base RuneLite updates, will my compiled version then no longer work?

File: runelite-client/src/main/java/net/runelite/client/plugins/gpu/GpuPlugin.java [hunk 1]
```diff
@@ -315,6 +321,13 @@ protected void shutDown()
 			client.setGpu(false);
 			client.setDrawCallbacks(null);
 
+			if (jawtWindow.getLock().getHoldCount() != NativeSurface.LOCK_SURFACE_NOT_READY)
+			{
+				jawtWindow.lockSurface();
+			}
+
+			glContext.destroy();
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/gpu/GpuPlugin.java]: This should probably be at the very end? Considering at startup it is before program/vao/texture alloc etc

**Comment by half-cambodian-hacker-man** [comment]
Works for me on Arch, AMD A8-7650K, GTX 960 with proprietary drivers on X11.

**Comment by l2-** [comment]
can someone run master with this patch applied and check if it still behaves as without this patch?
patch: https://pastebin.com/3tf3yCjc
to apply the patch, copy paste the contents of the pastebin above into a text file and do `git apply {textfile}`

**Comment by jaggerestep** [comment]
Can also confirm that it works for me on Ubuntu 16.04

GTX 1070 Driver Version: 396.54
Ryzen 1600 @ 3.75GHZ

**Pull request merged**